sudo/lib/system/term/ops.rs

```diff
@@ -0,0 +1,186 @@
+        },
+        sa_flags: 0,
+        sa_restorer: None,
+    };
+    // Reset `GOT_SIGTTOU`.
+    GOT_SIGTTOU.store(false, Ordering::SeqCst);
+    // Set `action` as the action for `SIGTTOU` and store the original action in `original_action`
+    // to restore it later.
+    unsafe { sigaction(SIGTTOU, &action, original_action.as_mut_ptr()) };
+    // Call `tcsetattr` until it suceeds and ignore interruptions if we did not receive `SIGTTOU`.
+    loop {
+        match cerr(unsafe { tcsetattr(fd, flags, tp) }) {
+            Ok(_) => break,
+            Err(err) => {
+                let got_sigttou = GOT_SIGTTOU.load(Ordering::SeqCst);
+                if got_sigttou || err.kind() != io::ErrorKind::Interrupted {
+                    return Err(err);
+                }
+            }
+        }
+    }
+    // Restore the original action.
+    unsafe { sigaction(SIGTTOU, original_action.as_ptr(), std::ptr::null_mut()) };
+
+    Ok(())
+}
+
+/// Copy the settings of the `src` terminal to the `dst` terminal.
+pub fn term_copy<S: AsRawFd, D: AsRawFd>(src: &S, dst: &D) -> io::Result<()> {
+    let src = src.as_raw_fd();
+    let dst = dst.as_raw_fd();
+
+    let mut tt_src = MaybeUninit::<termios>::uninit();
+    let mut tt_dst = MaybeUninit::<termios>::uninit();
+    let mut wsize = MaybeUninit::<winsize>::uninit();
+
+    cerr(unsafe { tcgetattr(src, tt_src.as_mut_ptr()) })?;
+    cerr(unsafe { tcgetattr(dst, tt_dst.as_mut_ptr()) })?;
+
+    let tt_src = unsafe { tt_src.assume_init() };
+    let mut tt_dst = unsafe { tt_dst.assume_init() };
+
+    // Clear select input, output, control and local flags.
+    tt_dst.c_iflag &= !INPUT_FLAGS;
+    tt_dst.c_oflag &= !OUTPUT_FLAGS;
+    tt_dst.c_cflag &= !CONTROL_FLAGS;
+    tt_dst.c_lflag &= !LOCAL_FLAGS;
+
+    // Copy select input, output, control and local flags.
+    tt_dst.c_iflag |= tt_src.c_iflag & INPUT_FLAGS;
+    tt_dst.c_oflag |= tt_src.c_oflag & OUTPUT_FLAGS;
+    tt_dst.c_cflag |= tt_src.c_cflag & CONTROL_FLAGS;
+    tt_dst.c_lflag |= tt_src.c_lflag & LOCAL_FLAGS;
+
+    // Copy special chars from src verbatim.
+    tt_dst.c_cc.copy_from_slice(&tt_src.c_cc);
+
+    // Copy speed from src (zero output speed closes the connection).
+    {
+        let mut speed = unsafe { cfgetospeed(&tt_src) };
+        if speed == libc::B0 {
+            speed = libc::B38400;
+            unsafe { cfsetospeed(&mut tt_dst, speed) };
+            speed = unsafe { cfgetispeed(&tt_src) };
+            unsafe { cfsetispeed(&mut tt_dst, speed) };
+        }
+    }
+
+    tcsetattr_nobg(dst, TCSAFLUSH, &tt_dst)?;
+
+    cerr(unsafe { ioctl(src, TIOCGWINSZ, &mut wsize) })?;
+    cerr(unsafe { ioctl(dst, TIOCSWINSZ, &wsize) })?;
+
+    Ok(())
+}
+
+/// Set the `fd` terminal to raw mode. Enable terminal signals if `with_signals` is set to `true`.  
+pub fn term_raw<F: AsRawFd>(fd: &F, with_signals: bool) -> io::Result<()> {
+    let fd = fd.as_raw_fd();
+
+    if !unsafe { CHANGED } {
```

Comment:
regarding these static variables, it's difficult to provide advice without knowing in what kind of context they'll be used. at the very least, I think it should be fine to change the type of `CHANGED` to `AtomicBool` and use Acq loads and Rel stores. even with that change I don't think this function will be reentrant-safe or signal-safe

Reply:
This particular function is used while resuming the terminal if `SIGCONT` arrives, to suspend the main sudo process if the command received `SIGTTOU` or `SIGTTIN`, and to set `/dev/tty` to raw mode before spawning the monitor if `sudo` was not called in a pipeline or in background mode. None of these happen inside a signal handler fortunately so the signal safety part doesn't worry me that much.

The reentrant-safe part is trickier. Given that we process the polled file descriptors sequentially, we shouldn't handle two signals at the same time twice from the parent, we might be safe but I'm not absolutely sure.